vizro-core/src/vizro/models/_components/card.py

```diff
@@ -1,48 +1,94 @@
-    )
+    text: str = Field("", description="Markdown string to create card text that should adhere to the CommonMark Spec.")
     href: str = Field(
         "",
         description="URL (relative or absolute) to navigate to. If not provided the Card serves as a text card only.",
     )
+    figure: Optional[CapturedCallable] = Field(
+        None, import_path=vc, description="Dynamic card function to be visualized on dashboard."
+    )
+
+    # Component properties for actions and interactions
+    _output_component_property: str = PrivateAttr("children")
+
+    # Validators
+    @validator("figure")
+    def validate_figure(cls, figure, values):
+        if figure is None:
+            return (
+                vc.nav_card(text=values["text"], href=values["href"], data_frame=pd.DataFrame())
+                if values["href"]
+                else vc.text_card(text=values["text"], data_frame=pd.DataFrame())
+            )
+        return figure
+
+    _validate_callable_mode = _callable_mode_validator_factory("card")
+    _validate_callable = validator("figure", allow_reuse=True, always=True)(_process_callable_data_frame)
+
+    def __call__(self, **kwargs):
+        kwargs.setdefault("data_frame", data_manager[self["data_frame"]].load())
+        figure = self.figure(**kwargs) if self.figure else None
+        return figure
+
+    def __getitem__(self, arg_name: str):
+        # pydantic discriminated union validation seems to try Graph["type"], which throws an error unless we
+        # explicitly redirect it to the correct attribute.
+        if self.figure:
+            if arg_name == "type":
+                return self.type
+            return self.figure[arg_name]
 
     @_log_call
     def build(self):
-        text = dcc.Markdown(self.text, dangerously_allow_html=False, id=self.id)
-        card_content = (
-            dbc.NavLink(
-                text,
-                href=get_relative_path(self.href) if self.href.startswith("/") else self.href,
-            )
-            if self.href
-            else text
-        )
+        # LQ: Don't really like this - is there a better way to do this? I need to provide different classNames to the outer
+        # dbc.Card based on the CapturedCallable provided. This is a bit hacky.
+        function_name = self.figure._function.__name__
 
-        card_class = "card-nav" if self.href else ""
-        return dbc.Card(card_content, className=card_class)
+        # LQ: Previously discussed to insert dbc.Card directly here and provide self.id to the outer container.
+        # However, this can affect several custom actions that rely on the id being previously provided to the inner
+        # text component instead of the outer card component. Shall we keep as is or pass the id through the relevant inner
+        # components? The last would require to provide an id argument to each CapturedCallable, which is also not optimal.
```

Comment:
What I am referring to here is that previously, we've been providing the `id` to the `Card.text` as we've said that this is the reactive sub-component, and a user might want to update it.

The logic with the `_input_component_id` always appends it to the `figure`, which is the outer container of the Card and not the `text`. I think that also makes more sense, because given that anyone can provide a CapturedCallable with a different card component hierarchy, it's safer to say the entire content needs to be updated.

However, if we implement this change, I think it might break for people who have already been using the `Card.text` for some custom actions because now the `id` is not provided to the `Card.text` anymore but to the entire `Card`. So, all the content needs to be updated.

Reply:
If we are fine that the `id` is now provided to the outer container the dbc.Card itself, I think the current implementation would also work even without `_input_component_id` 🤔